src/main.rs

```diff
@@ -228,152 +157,115 @@ async fn main() -> Result<(), anyhow::Error> {
         }
     };
 
-    // Install package
-    info!("Installing to: '{}'", install_path);
-    extract(&pkg_path, pkg_fmt, &install_path)?;
+    debug!("Using install path: {}", install_path.display());
 
-    
-    info!("Installation done!");
+    info!("Downloading package from: '{}'", rendered);
 
-    Ok(())
-}
+    // Download package
+    let pkg_path = temp_dir.path().join(format!("pkg-{}.{}", pkg_name, pkg_fmt));
+    download(&rendered, pkg_path.to_str().unwrap()).await?;
 
-/// Download a file from the provided URL to the provided path
-async fn download<P: AsRef<Path>>(url: &str, path: P) -> Result<(), anyhow::Error> {
+    #[cfg(incomplete)]
+    {
+        // Fetch and check package signature if available
+        if let Some(pub_key) = meta.as_ref().map(|m| m.pub_key.clone() ).flatten() {
+            debug!("Found public key: {}", pub_key);
 
-    debug!("Downloading from: '{}'", url);
+            // Generate signature file URL
+            let mut sig_ctx = ctx.clone();
+            sig_ctx.format = "sig".to_string();
+            let sig_url = sig_ctx.render(&pkg_url)?;
 
-    let resp = reqwest::get(url).await?;
+            debug!("Fetching signature file: {}", sig_url);
 
-    if !resp.status().is_success() {
-        error!("Download error: {}", resp.status());
-        return Err(anyhow::anyhow!(resp.status()));
-    }
+            // Download signature file
+            let sig_path = temp_dir.path().join(format!("{}.sig", pkg_name));
+            download(&sig_url, &sig_path).await?;
 
-    let bytes = resp.bytes().await?;
+            // TODO: do the signature check
+            unimplemented!()
 
-    debug!("Download OK, writing to file: '{:?}'", path.as_ref());
+        } else {
+            warn!("No public key found, package signature could not be validated");
+        }
+    }
 
-    std::fs::write(&path, bytes)?;
+    // Extract files
+    let bin_path = temp_dir.path().join(format!("bin-{}", pkg_name));
+    extract(&pkg_path, pkg_fmt, &bin_path)?;
 
-    Ok(())
-}
-
-fn extract<S: AsRef<Path>, P: AsRef<Path>>(source: S, fmt: PkgFmt, path: P) -> Result<(), anyhow::Error> {
-    match fmt {
-        PkgFmt::Tar => {
-            // Extract to install dir
-            debug!("Extracting from archive '{:?}' to `{:?}`", source.as_ref(), path.as_ref());
+    // Bypass cleanup if disabled
+    if opts.no_cleanup {
+        let _ = temp_dir.into_path();
+    }
 
-            let dat = std::fs::File::open(source)?;
-            let mut tar = Archive::new(dat);
+    // List files to be installed
+    // TODO: check extracted files are sensible / filter by allowed files
+    // TODO: this seems overcomplicated / should be able to be simplified?
+    let bin_files = std::fs::read_dir(&bin_path)?;
+    let bin_files: Vec<_> = bin_files.filter_map(|f| f.ok() ).map(|f| {
```

Comment:
this logic a) needs to be torn out into the library (with tests) and b) is definitely broken for windows (`*.exe`)